listener: ensure `getBalancedHandlerByAddress` works with ipv4_compat

soulxu opened the pull request:
Commit Message: listener: ensure `getBalancedHandlerByAddress` works with ipv4_compat
Additional Description:
When a listener is listening on ipv6 any-address or ipv4-compatible ipv6 address, and with ipv4_compat flag as true, it is expected to accept the request both from ipv4 and ipv6 address.

This PR will add a duplicated map item for ipv4 any-address or ipv4 compatible address if the listener has ipv4_compat flag, that will ensure `getBalancedHandlerByAddress` can match ipv4 request for it.

Also, add a more strict check on `ipv4_compat` flag. The `ip4_compat` only can be set as true when the address is Ipv6 any address or Ipv4-mapped Ipv6 address. The runtime flag is added `envoy.reloadable_features.strict_check_on_ipv4_compat` and the default is true.

Risk Level: low
Testing: unittest 
Docs Changes: n/a
Release Notes: A note added to minor incompatible change.
Platform Specific Features: n/a
Fixes #18134

Signed-off-by: He Jie Xu <hejie.xu@intel.com>

jacob-delgado reviewed (commented):
Similar to what I had, but I couldn't get the tests in a good state. Thanks for doing this!

jacob-delgado commented:
cc @soulxu and @ggreenway since this bug was written for v1.19+ should it be backported? I can help backport it if @soulxu is busy. Let me know as I believe you triage bugs for back porting. Thanks!

ggreenway commented:
@lambdai can you take a first pass on this?

lambdai commented:
> @lambdai can you take a first pass on this?

ack

soulxu commented:
@lambdai thanks

lambdai reviewed (changes requested):
Other than the inline commentsat connection_handler, I have below questions 

1. Is `[::]` w/o ipv6only actually listen on ipv4 any address? The listen and bind is executed on main thread, but I don't see the change there. 
  If not, the current code only fix the part that finding listener not binding_to_port, but not the general re balancing among physical listener.

The partial fix makes it hard to deduct tcp_listener_map_by_address_. Perhaps it's ok to leave a TODO

soulxu commented on source/server/connection_handler_impl.cc:
```diff
@@ -94,6 +94,16 @@ void ConnectionHandlerImpl::addListener(absl::optional<uint64_t> overridden_list
+    auto& address = details->address_;
+    // If the address is IPv6 and isn't v6only, adds an extra ipv4 any-address item to the map.
+    // Then this allows the `getBalancedHandlerByAddress` can match the ipv4 any-address also.
+    if (address->type() == Network::Address::Type::Ip &&
+        address->ip()->version() == Network::Address::IpVersion::v6 &&
+        !address->ip()->ipv6()->v6only()) {
```
After looking https://datatracker.ietf.org/doc/html/rfc3493#section-3.7, I see now. you are right, it is more than just any address.  I should get the compatible ipv4 address out of the ipv6, and put it into the map.

lambdai replied:
Thank you! This is awesome.

soulxu commented:
> Other than the inline commentsat connection_handler, I have below questions
> 
> 1. Is `[::]` w/o ipv6only actually listen on ipv4 any address? The listen and bind is executed on main thread, but I don't see the change there.

Yes, I tested that, the  `[::]` listen on ipv4 any address also. That should be done by the kernel.

lambdai reviewed (changes requested):
Thank you for the update!
LGTM modulo the exception throw.

soulxu commented:
> Thank you for the update! LGTM modulo the exception throw.

thanks for the review! Do you want to remove the exception?

soulxu commented:
/retest

soulxu commented:
/wait

soulxu commented:
after conn handler fixed, this is ready for review again.

soulxu commented on source/server/connection_handler_impl.cc:
```diff
@@ -94,6 +94,28 @@ void ConnectionHandlerImpl::addListener(absl::optional<uint64_t> overridden_list
+        !address->ip()->ipv6()->v6only()) {
+      if (address->ip()->isAnyAddress()) {
+        tcp_listener_map_by_address_.insert_or_assign(
+            Network::Utility::getIpv4AnyAddress(address->ip()->port())->asStringView(), details);
```
I need update this.

soulxu commented:
/wait

soulxu commented:
success to run istio integration test locally

lambdai commented:
> after conn handler fixed, this is ready for review again.

Thank you for the fix and holding this.

Can you remove the exception and move to listener manager? See the inline comment.

Also, can you explain how the listener manager handles two listener config: one is ipv4-any:80 and another is ipv6-any:80?

To be more precise, I think kernel allows the above 2 listening socket, but my hunch is that tcp_listener_map_by_address_ doesn't like two ipv4-any

soulxu commented:
> > after conn handler fixed, this is ready for review again.
> 
> Thank you for the fix and holding this.
> 
> Can you remove the exception and move to listener manager? See the inline comment.

Got it, I moved to ListenerImpl. But in the end, I realize this is API behavior change. So I added a runtime flag this also.

The old behavior is you can set `ipv4_compat` flag on any kind of address, and linux kernel won't complain that also. But actually, only ipv6 any address and ipv4-mapped ipv6 address can work. From the API point of view, this seems wrong, since it is silent on something that doesn't work. But yes, this will break some people if they didn't use the `ipv4_compat` correctly before.  Anyway, let me know if this is the right choice.

> 
> Also, can you explain how the listener manager handles two listener config: one is ipv4-any:80 and another is ipv6-any:80?
> 
> To be more precise, I think kernel allows the above 2 listening socket, but my hunch is that tcp_listener_map_by_address_ doesn't like two ipv4-any

The ipv4 any address is "0.0.0.0" and the ipv6 any address is "::". And the tcp_listener_map_by_address_  is using those address string as key. so it should be ok. Although we have multiple way to right any address in Ipv6, but we have code to make it as canonical format https://github.com/envoyproxy/envoy/blob/main/source/common/network/address_impl.cc#L261

soulxu commented:
> 
> > Also, can you explain how the listener manager handles two listener config: one is ipv4-any:80 and another is ipv6-any:80?
> 
> The listener manager can distinguish the two address by
> 
> https://github.com/envoyproxy/envoy/blob/027d6ca15adeae8cd478fcce67fc24322318e52e/source/common/network/address_impl.cc#L268-L272
> 
> The dynamic cast should return nullptr when it is same ip version, so listener manager is ok here.
> 
> > To be more precise, I think kernel allows the above 2 listening socket, but my hunch is that tcp_listener_map_by_address_ doesn't like two ipv4-any
> 
> The ipv4 any address is "0.0.0.0" and the ipv6 any address is "::". And the tcp_listener_map_by_address_ is using those address string as key. so it should be ok. Although we have multiple way to right any address in Ipv6, but we have code to make it as canonical format https://github.com/envoyproxy/envoy/blob/main/source/common/network/address_impl.cc#L261

Thanks for this good call. In the end, I found a corner case. We can one listener for "::" and ipv4_compat, another listener is for "0.0.0.0". The kernel will route the request to "0.0.0.0". And same behavior on ipv4-mapped ipv6 address.

ggreenway commented on source/server/connection_handler_impl.cc:
```diff
@@ -95,6 +95,29 @@ void ConnectionHandlerImpl::addListener(absl::optional<uint64_t> overridden_list
+            Network::Address::Ipv4Instance(address->ip()->port()).asStringView(), details);
+      } else {
+        auto v4_compatible_addr = address->ip()->ipv6()->v4CompatibleAddress();
+        // Remove this check when runtime flag
```
It may be better to leave the check in, and trigger an ENVOY_BUG when it fails, after the runtime check is removed. That seems better than crashing, and this check is inexpensive.

soulxu replied:
When `envoy.reloadable_features.strict_check_on_ipv4_compat` is false, should we go back to the original behavior instead of let Envoy stop?

mattklein123 replied:
ENVOY_BUG won't stop in a production build, so I think it might be worth it to do it in this case.

soulxu replied:
I see, that makes sense, thanks!

ggreenway commented on source/server/connection_handler_impl.cc:
```diff
@@ -95,6 +95,29 @@ void ConnectionHandlerImpl::addListener(absl::optional<uint64_t> overridden_list
+        // Remove this check when runtime flag
+        // `envoy.reloadable_features.strict_check_on_ipv4_compat` deprecated.
+        if (v4_compatible_addr.ok()) {
+          // Since both "::" with ipv4_compat and "0.0.0.0" can be supported.
```
Is it a valid config to have listeners on both :: and 0.0.0.0? What is the expected behavior? Is it OS dependent?

soulxu replied:
Replied in the main thread.

soulxu replied:
oops, I should use `insert_or_assign` for any address also, sorry for I just realize that.

ggreenway commented on source/server/listener_impl.cc:
```diff
@@ -348,6 +348,19 @@ ListenerImpl::ListenerImpl(const envoy::config::listener::v3::Listener& config,
+      Runtime::runtimeFeatureEnabled("envoy.reloadable_features.strict_check_on_ipv4_compat")) {
+    if (address_->ip()->version() == Network::Address::IpVersion::v6) {
+      if (!address_->ip()->ipv6()->v4CompatibleAddress().ok()) {
+        throw EnvoyException("Only the Ipv6 any address and Ipv4-mapped Ipv6 address can be set "
```
I think it may be clearer to say `::` and `0.0.0.0` instead of `any`

ggreenway replied:
`IPv6` and `IPv4` (capitalization) for consistency with other docs/messages.

ggreenway replied:
`Only IPv6 address '::' or valid IPv4-mapped IPv6 addresses can set ipv4_compat`

soulxu replied:
got it.

ggreenway reviewed (changes requested):
I'm concerned about what happens if there are multiple overlapping wildcard listeners with ipv4 and ipv6+ipv4_compat. Can you elaborate on whether that config is even possible, what the expected behavior is, etc?

/wait

soulxu commented:
> I'm concerned about what happens if there are multiple overlapping wildcard listeners with ipv4 and ipv6+ipv4_compat. Can you elaborate on whether that config is even possible, what the expected behavior is, etc?
> 
> /wait

Envoy supports that one listener is listening on `::` with `ipv4_compat` flag, and another listener is listening on `0.0.0.0`.  The ipv4 request is going to the second listener and the ipv6 request is going to the first listener. This is the Linux behavior. I don't know other OS behavior yet, I should be able to test it on Windows when I got the chance. For now I matched this behavior for the redirect in the code. https://github.com/envoyproxy/envoy/pull/19716/files#diff-3df0a7bd0288faa35b611e2bc11ea94713d9ca617ad16dfab81386c814f5e1eeR1554

But I do a test today on the ipv4-mapped ipv6 address. It is surprising that it has different behavior. We can both listen on `::FFFF:192.168.0.11 and `192.168.0.1`. But all the requests are going to the `192.168.0.1` listener even for ipv6 request. Using `netstat`, I can see envoy is only listening on "192.168.0.1".

I wrote a simple socket program to listen on those two addresses, it was rejected by kernel, only one of addresses can listen. In the end, I can listen to two addresses with `SO_REUSEADDR` opt. Also, interesting thing is that Envoy can listen on both addresses with `enable_reuse_port` as false, but I found this https://github.com/envoyproxy/envoy/blob/d1f5c2bc389d6217bca008c82b6a49019025f2d3/source/common/network/listen_socket_impl.h#L125-L133
That is probably the reason why Envoy always can listen to both addresses.

This is also only the Linux behavior.

But seems we should reject the user create two listeners for "::FFFF:192.168.0.1" and "192.168.0.1" at same time, it seems meanless.

I'm not sure on the use-case of "::" with ipv4_compat and "0.0.0.0" are valid or not, at least the linux works with that. I can keep matching that behavior for the redirect.

soulxu commented on source/common/network/address_impl.cc:
```diff
@@ -233,6 +237,14 @@ std::string Ipv6Instance::Ipv6Helper::makeFriendlyAddress() const {
+    ipv6ToIpv4CompatibleAddress(&address_, &sin);
+    return Address::InstanceFactory::createInstancePtr<Address::Ipv4Instance>(&sin);
+  }
+  return absl::InvalidArgumentError(fmt::format("It isn't Ipv4-mapped Ipv6 address"));
```
got it.

soulxu commented on source/server/connection_handler_impl.cc:
```diff
@@ -95,6 +95,35 @@ void ConnectionHandlerImpl::addListener(absl::optional<uint64_t> overridden_list
+        auto ipv4_any_listener = tcp_listener_map_by_address_.find(
+            Network::Address::Ipv4Instance(address->ip()->port()).asStringView());
+        if (ipv4_any_listener == tcp_listener_map_by_address_.end() ||
+            ipv4_any_listener->second->listener_->listener() == nullptr) {
```
Take care of one more case. There could be one "0.0.0.0" listener was stopped, then a "::" with ipv4_compat listener was added, then ensure the ipv6 listener will replace the ipv4 one.

soulxu commented:
> > But I do a test today on the ipv4-mapped ipv6 address. It is surprising that it has different behavior. We can both listen on `::FFFF:192.168.0.11 and `192.168.0.1`. But all the requests are going to the `192.168.0.1`listener even for ipv6 request. Using`netstat`, I can see envoy is only listening on "192.168.0.1". @lambdai sorry, I didn't find this before, probably I mess with test.
> 
> @lambdai @ggreenway If we are going to reject this from the API, it seems a separated PR. So I filed another issue for this in case this needs a broader discussion #19981

I removed the part of taking care of both ipv4 and ipv4-mapped ipv6 are listening since both linux and windows don't work with it. 

But still keep the part of both "0.0.0.0" and "::" are listening, but I will update if we don't want to support that also.

soulxu commented:
> > > But I do a test today on the ipv4-mapped ipv6 address. It is surprising that it has different behavior. We can both listen on `::FFFF:192.168.0.11 and `192.168.0.1`. But all the requests are going to the `192.168.0.1`listener even for ipv6 request. Using`netstat`, I can see envoy is only listening on "192.168.0.1". @lambdai sorry, I didn't find this before, probably I mess with test.
> > 
> > 
> > @lambdai @ggreenway If we are going to reject this from the API, it seems a separated PR. So I filed another issue for this in case this needs a broader discussion #19981
> 
> I removed the part of taking care of both ipv4 and ipv4-mapped ipv6 are listening since both linux and windows don't work with it.
> 

Clarify this. Above is about two listeners are set `bind_to_port` as `true`. After a test, using the `original dst`, the kernel only return ipv4 address even request with ipv4-mapped address. So two listeners (ipv4 and ipv4-mapped) with `bind_to_port` as `false`, seems useless also.

lambdai commented:
@soulxu 
Thanks for the update. Your latest description matches my observation.

In the kernel ::FFFF:192.168.0.11 address requires ipv4 stack.

In this way, envoy should actively reject "::FFFF:192.168.0.11" if 192.168.0.11 listener exists. I think it could be in separate PR

lambdai commented on source/server/connection_handler_impl.cc:
```diff
@@ -95,6 +95,36 @@ void ConnectionHandlerImpl::addListener(absl::optional<uint64_t> overridden_list
+        auto ipv4_any_listener = tcp_listener_map_by_address_.find(ipv4_any_address);
+        if (ipv4_any_listener == tcp_listener_map_by_address_.end() ||
+            ipv4_any_listener->second->listener_->listener() == nullptr) {
+          tcp_listener_map_by_address_.insert_or_assign(ipv4_any_address, details);
```
Can you guard this with the runtime behavior? 

My concern is that the new matching behavior, though correct, will break a lot of existing [::] listener configs

because the new behavior requires **both** the `filter chains` ipv4 set and ipv6 set. I don't think all the control planes are that strong.

CC @howardjohn

soulxu replied:
Emm...this is the regression https://github.com/envoyproxy/envoy/issues/18134 we are trying to fix. So...it should be a regression, not a behavior change.

lambdai replied:
@soulxu  Make sense!

soulxu commented:
> @soulxu Thanks for the update. Your latest description matches my observation.
> 
> In the kernel ::FFFF:192.168.0.11 address requires ipv4 stack.
> 
> In this way, envoy should actively reject "::FFFF:192.168.0.11" if 192.168.0.11 listener exists. I think it could be in separate PR

thanks for the feedback! I will reject them by separated PR, double-check this rejection includes the bind_to_port=false case also, right?

soulxu commented on source/server/connection_handler_impl.cc:
```diff
@@ -95,6 +95,28 @@ void ConnectionHandlerImpl::addListener(absl::optional<uint64_t> overridden_list
+      } else {
+        auto v4_compatible_addr = address->ip()->ipv6()->v4CompatibleAddress();
+        if (!v4_compatible_addr.ok() && !v4_compatible_addr.status().ok()) {
+          throw EnvoyException(fmt::format("Can't get v4 compatible address for {}: {}",
```
Yea, I already moved to ListenerImpl. I guess this is your old comment which forgets to submit.

lambdai commented:
re: bind_to_port

+1 to make the discrepancy as little as possible.

lambdai reviewed (approved):
LGTM!

lambdai commented:
/retest

lambdai commented:
CC @ggreenway for the next step

rojkov commented:
@ggreenway gentle ping

mattklein123 commented on envoy/network/address.h:
```diff
@@ -50,6 +55,11 @@ class Ipv6 {
   virtual bool v6only() const PURE;
+
+  /**
+   * @return Ipv4 address from Ipv4-compatible Ipv6 address.
```
Presumably this fails if the v6 address is not in the v4 compatible range? Can you make that more clear in the comments?

soulxu replied:
Yes, will update

mattklein123 commented on source/common/network/address_impl.cc:
```diff
@@ -233,6 +237,14 @@ std::string Ipv6Instance::Ipv6Helper::makeFriendlyAddress() const {
+    ipv6ToIpv4CompatibleAddress(&address_, &sin);
+    return Address::InstanceFactory::createInstancePtr<Address::Ipv4Instance>(&sin);
+  }
+  return absl::InvalidArgumentError("It isn't Ipv4-mapped Ipv6 address");
```
Instead of using StatusOr perhaps just return nullptr? I'm not sure the status provides much value here?

soulxu replied:
yes, agree with you, will update

mattklein123 commented on source/common/network/address_impl.cc:
```diff
@@ -61,16 +73,8 @@ StatusOr<Address::InstanceConstSharedPtr> addressFromSockAddr(const sockaddr_sto
     const struct sockaddr_in6* sin6 = reinterpret_cast<const struct sockaddr_in6*>(&ss);
     ASSERT(AF_INET6 == sin6->sin6_family);
     if (!v6only && IN6_IS_ADDR_V4MAPPED(&sin6->sin6_addr)) {
```
I think you can just call v4CompatibleAddress() here and if null do the other branch?

soulxu replied:
Emm...the Instance obj doesn't create yet. We create different Instance obj below these few lines and depend on this check result.

https://github.com/envoyproxy/envoy/pull/19716/files/228cc7e14615516b1397e3501f8dfa66cc5676a1#diff-6925fefb8a60a68e4059b0d4fdbb943c4b3af9fa77df231724cabcc1f1b35f9fL74-L76

mattklein123 commented on source/common/network/address_impl.cc:
```diff
@@ -233,6 +237,14 @@ std::string Ipv6Instance::Ipv6Helper::makeFriendlyAddress() const {
   ASSERT(str == ptr);
   return ptr;
 }
+StatusOr<InstanceConstSharedPtr> Ipv6Instance::Ipv6Helper::v4CompatibleAddress() const {
```
nit: newline above this line

mattklein123 commented on source/common/network/address_impl.cc:
```diff
@@ -260,6 +272,7 @@ Ipv6Instance::Ipv6Instance(const std::string& address, uint32_t port,
   // Just in case address is in a non-canonical format, format from network address.
   ip_.friendly_address_ = ip_.ipv6_.makeFriendlyAddress();
   friendly_name_ = fmt::format("[{}]:{}", ip_.friendly_address_, ip_.port());
+  ip_.ipv6_.v6only_ = v6only;
```
Not your code, but is there any reason this part of the function can't reuse initHelper()? (I didn't look carefully just curious, seems like some duplication.)

mattklein123 replied:
Sure that's fine. No worries.

mattklein123 reviewed (changes requested):
Thanks this makes sense to me. Just some small comments.

/wait

soulxu commented:
> Thanks this makes sense to me. Just some small comments.
> 
> /wait

Thanks for the review!

soulxu commented on source/server/connection_handler_impl.cc:
```diff
@@ -115,12 +115,11 @@ void ConnectionHandlerImpl::addListener(absl::optional<uint64_t> overridden_list
           tcp_listener_map_by_address_.insert_or_assign(v4_compatible_addr->asStringView(),
                                                         details);
+        } else {
+          IS_ENVOY_BUG("If the Ipv6 address isn't v6only, it supposed to be Ipv4-mapped");
```
Yes, config load will block it once runtime flag gone. 
https://github.com/envoyproxy/envoy/pull/19716/files#diff-b0d5e1a1317df8e22a556ee74d6f728379c5f85ce640a9404acc14bfe8192ad3R356-R357

How a about a warning near the exception. And the warning will gone with runtime flag in the future.

mattklein123 reviewed (approved):
Thanks!

moderation commented:
@soulxu  I'm having trouble with one of my configs now.
```json
                 "address": {
                     "socket_address": {
                         "address": "::0",
                         "ipv4_compat": true,
                         "port_value": 443
                     }
                 },
```
errors with
```
[2022-03-03 21:37:30.259][183239][critical][main] [source/server/server.cc:117] error initializing configuration 'sync-front-envoy-v3.json': Only IPv6 address '::' or valid IPv4-mapped IPv6 address can set ipv4_compat: [::]:443
[2022-03-03 21:37:30.259][183239][info][main] [source/server/server.cc:933] exiting
Only IPv6 address '::' or valid IPv4-mapped IPv6 address can set ipv4_compat: [::]:443
```
Do I have the config incorrect?

soulxu commented:
> @soulxu I'm having trouble with one of my configs now.
> 
> ```json
>                  "address": {
>                      "socket_address": {
>                          "address": "::0",
>                          "ipv4_compat": true,
>                          "port_value": 443
>                      }
>                  },
> ```
> 
> errors with
> 
> ```
> [2022-03-03 21:37:30.259][183239][critical][main] [source/server/server.cc:117] error initializing configuration 'sync-front-envoy-v3.json': Only IPv6 address '::' or valid IPv4-mapped IPv6 address can set ipv4_compat: [::]:443
> [2022-03-03 21:37:30.259][183239][info][main] [source/server/server.cc:933] exiting
> Only IPv6 address '::' or valid IPv4-mapped IPv6 address can set ipv4_compat: [::]:443
> ```
> 
> Do I have the config incorrect?

Yes, you have the config correct.

Sorry for I made stupid mistake, I forget to add a check for any address

https://github.com/envoyproxy/envoy/pull/19716/files#diff-b0d5e1a1317df8e22a556ee74d6f728379c5f85ce640a9404acc14bfe8192ad3R351-R354

Even I have an error message for `::` is valid.

soulxu commented:
> > @soulxu I'm having trouble with one of my configs now.
> > ```json
> >                  "address": {
> >                      "socket_address": {
> >                          "address": "::0",
> >                          "ipv4_compat": true,
> >                          "port_value": 443
> >                      }
> >                  },
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > errors with
> > ```
> > [2022-03-03 21:37:30.259][183239][critical][main] [source/server/server.cc:117] error initializing configuration 'sync-front-envoy-v3.json': Only IPv6 address '::' or valid IPv4-mapped IPv6 address can set ipv4_compat: [::]:443
> > [2022-03-03 21:37:30.259][183239][info][main] [source/server/server.cc:933] exiting
> > Only IPv6 address '::' or valid IPv4-mapped IPv6 address can set ipv4_compat: [::]:443
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Do I have the config incorrect?
> 
> Yes, you have the config correct.
> 
> Sorry for I made stupid mistake, I forget to add a check for any address
> 
> https://github.com/envoyproxy/envoy/pull/19716/files#diff-b0d5e1a1317df8e22a556ee74d6f728379c5f85ce640a9404acc14bfe8192ad3R351-R354
> 
> Even I have an error message for `::` is valid.

@moderation here is the fix https://github.com/envoyproxy/envoy/pull/20214

Sorry again
